sklearn/cluster/_kmeans.py

```diff
@@ -1665,160 +1751,119 @@ def fit(self, X, y=None, sample_weight=None):
-            if _mini_batch_convergence(
-                    self, iteration_idx, n_iter, tol, n_samples,
-                    centers_squared_diff, batch_inertia, convergence_context,
-                    verbose=self.verbose):
-                break
+        centers = init_centers
+        centers_new = np.empty_like(centers)
+
+        # Initialize counts
+        self._counts = np.zeros(self.n_clusters, dtype=X.dtype)
+
+        # Attributes to monitor the convergence
+        self._ewa_inertia = None
+        self._ewa_inertia_min = None
+        self._no_improvement = 0
+
+        # Initialize number of samples seen since last reassignment
+        self._n_since_last_reassign = 0
+
+        n_steps_per_epoch = int(np.ceil(n_samples / self._batch_size))
+        n_steps = self.max_iter * n_steps_per_epoch
+
+        with threadpool_limits(limits=1, user_api="blas"):
+            # Perform the iterative optimization until convergence
+            for i in range(n_steps):
+                # Sample a minibatch from the full dataset
+                minibatch_indices = random_state.randint(0, n_samples,
+                                                         self._batch_size)
+
+                # Perform the actual update step on the minibatch data
+                batch_inertia = _mini_batch_step(
+                    X=X[minibatch_indices],
+                    x_squared_norms=x_squared_norms[minibatch_indices],
+                    sample_weight=sample_weight[minibatch_indices],
+                    centers=centers,
+                    centers_new=centers_new,
+                    weight_sums=self._counts,
+                    random_state=random_state,
+                    random_reassign=self._random_reassign(),
+                    reassignment_ratio=self.reassignment_ratio,
+                    verbose=self.verbose,
+                    n_threads=self._n_threads)
+
+                if self._tol > 0.0:
+                    centers_squared_diff = np.sum((centers_new - centers)**2)
+                else:
+                    centers_squared_diff = 0
+
+                centers, centers_new = centers_new, centers
+
+                # Monitor convergence and do early stopping if necessary
+                if self._mini_batch_convergence(
+                        i, n_steps, n_samples, centers_squared_diff,
+                        batch_inertia):
+                    break
+
+        self.cluster_centers_ = centers
 
-        self.n_iter_ = iteration_idx + 1
+        self.n_steps_ = i + 1
+        self.n_iter_ = np.ceil((i + 1) / n_steps_per_epoch)
 
         if self.compute_labels:
-            self.labels_, self.inertia_ = \
-                    self._labels_inertia_minibatch(X, sample_weight)
+            self.labels_, self.inertia_ = _labels_inertia_threadpool_limit(
+                X, sample_weight, x_squared_norms, self.cluster_centers_,
+                n_threads=self._n_threads)
```

Comment:
If `compute_labels` is `False`, one could still expose `_ewa_inertia` as a public `inertia_` attribute (with a disclaimer in the docstring that in this case the value if not exact.

Reply:
I'm not sure because it would not match the labels since _ewa_inertia is computed before updating the centers.
If one computes manually the labels on his batch he will find an inertia different from the reported inertia_

Reply:
It think it's ok as long as it's documented as an approximate value when `compute_labels=False`.

Having this as a public attribute might be interesting for very large problems on which won't don't want to waste time with `compute_labels` on the full training set but one would still like to have a check that the model converged to a good enough minimum for instance.

Reply:
I agree it's interesting to have that. Done. Notice that I return `self._ewa_inertia * n_samples` to be able to compare with the compute_labels=True case.